File: patterns/Factory.py
from abc import ABC, abstractmethod

# --- Product ---
class Pizza(ABC):
    @abstractmethod
    def prepare(self):
        pass

class NYStyleCheesePizza(Pizza):
    def prepare(self):
        return "Preparing NY-style cheese pizza"

class ChicagoStyleCheesePizza(Pizza):
    def prepare(self):
        return "Preparing Chicago-style cheese pizza"

# --- Creator ---
class PizzaStore(ABC):
    @abstractmethod
    def create_pizza(self, type_: str) -> Pizza:
        pass

    def order_pizza(self, type_: str):
        pizza = self.create_pizza(type_)
        return pizza.prepare()

# --- Concrete Creators ---
class NYPizzaStore(PizzaStore):
    def create_pizza(self, type_: str):
        if type_ == "cheese":
            return NYStyleCheesePizza()
        raise ValueError("Unknown pizza type")

class ChicagoPizzaStore(PizzaStore):
    def create_pizza(self, type_: str):
        if type_ == "cheese":
            return ChicagoStyleCheesePizza()
        raise ValueError("Unknown pizza type")

# --- Usage ---
store = NYPizzaStore()
print(store.order_pizza("cheese"))

File: patterns/Factory.js
// --- Product ---
class Pizza {
  prepare() {
    throw new Error("Must implement prepare()");
  }
}

class NYStyleCheesePizza extends Pizza {
  prepare() {
    return "Preparing NY-style cheese pizza";
  }
}

class ChicagoStyleCheesePizza extends Pizza {
  prepare() {
    return "Preparing Chicago-style cheese pizza";
  }
}

// --- Creator ---
class PizzaStore {
  orderPizza(type) {
    const pizza = this.createPizza(type);
    return pizza.prepare();
  }

  createPizza(type) {
    throw new Error("Subclass must implement createPizza()");
  }
}

// --- Concrete Creators ---
class NYPizzaStore extends PizzaStore {
  createPizza(type) {
    if (type === "cheese") return new NYStyleCheesePizza();
    throw new Error("Unknown pizza type");
  }
}

class ChicagoPizzaStore extends PizzaStore {
  createPizza(type) {
    if (type === "cheese") return new ChicagoStyleCheesePizza();
    throw new Error("Unknown pizza type");
  }
}

// --- Usage ---
const store = new NYPizzaStore();
console.log(store.orderPizza("cheese"));